From 841bbfcf64861953bc45a7b379f2c89524905408 Mon Sep 17 00:00:00 2001
From: Sathyanarayanan Kuppuswamy <sathyanarayanan.kuppuswamy@intel.com>
Date: Fri, 14 Dec 2012 15:26:13 -0800
Subject: [PORT FROM MAIN] WIFI[PATCH 2/2]: Move gpio resource allocation from platform code to wifi driver

BZ: 75109

Currently we are allocating wifi gpio resources in platform code. This
patches move it to driver.

Change-Id: Ia0384ed4b43f3d42ce408a309a6481a144c8261f
Orig-Change-Id: Id4a6f88b7d501bf7d551b6d51054abaae0f32fe8
Signed-off-by: Sathyanarayanan Kuppuswamy <sathyanarayanan.kuppuswamy@intel.com>
Signed-off-by: Aymen Zayet <aymen.zayet@intel.com>
Reviewed-on: http://android.intel.com:8080/81871
Reviewed-by: Yang, Fei <fei.yang@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>

diff --git a/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c b/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c
index c7505d0..9bb934a 100644
--- a/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c
+++ b/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c
@@ -230,6 +230,12 @@ static int __devinit wl1271_probe(struct sdio_func *func,
 		goto out_free_glue;
 	}
 
+	if (wlan_data->hw_init) {
+		ret = wlan_data->hw_init(wlan_data);
+		if (ret)
+			goto out_free_glue;
+	}
+
 	/* if sdio can keep power while host is suspended, enable wow */
 	mmcflags = sdio_get_host_pm_caps(func);
 	dev_dbg(glue->dev, "sdio PM caps = 0x%x\n", mmcflags);
@@ -292,12 +298,19 @@ out:
 static void __devexit wl1271_remove(struct sdio_func *func)
 {
 	struct wl12xx_sdio_glue *glue = sdio_get_drvdata(func);
-
+	struct wl12xx_platform_data *wlan_data;
+
+	wlan_data = wl12xx_get_platform_data();
+	BUG_ON(!wlan_data);
 	/* Undo decrement done above in wl1271_probe */
 	pm_runtime_get_noresume(&func->dev);
 
 	platform_device_del(glue->core);
 	platform_device_put(glue->core);
+	/* Free platform resources */
+	if (wlan_data->hw_deinit)
+		wlan_data->hw_deinit(wlan_data);
+
 	kfree(glue);
 }
 
